### PR TITLE
zombienet: warp-sync integration test added

### PR DESCRIPTION
**Pull request by michalkucharczyk**
- `warp-sync` integration test for substrate added (refer to `zombienet/0001-basic-warp-sync/0001-README.md`)
- publishing temporary substrate image added for PR (`[docker.io/paritypr/substrate](https://hub.docker.com/r/paritypr/substrate/tags)`),
- zombienet stage added to CI pipeline,

Fixes: #11471

### CODE REVIEW

**Comment by arkpar** [comment]
"Warp sync is complete" and block height is not really an indication of the warp completion. There's also a second phase, where block history is downloaded and imported.  It should end with "Block history download is complete." message and the node should have not just the latest block (12133), but previous blocks as well, i.e. 12132

**Comment by michalkucharczyk** [comment]
This PR is blocked by:
- https://github.com/paritytech/substrate/pull/11727
- https://github.com/paritytech/zombienet/issues/578
- https://github.com/paritytech/zombienet/issues/580

File: scripts/ci/gitlab/pipeline/publish.yml [hunk 1]
```diff
@@ -38,17 +39,50 @@
...
     PRODUCT:                       substrate
 
+publish-docker-substrate-temporary:
+  extends:                         .build-push-image-temporary
+  needs:
+    - job:                         build-linux-substrate
+      artifacts:                   true
+  variables:
+    PRODUCT:                       substrate
+    PRODUCT_ALIAS:                 substrate-debug
```
Comment (michalkucharczyk) [inline scripts/ci/gitlab/pipeline/publish.yml:77]: We could use [`docker.io/paritypr/substrate`](https://hub.docker.com/r/paritypr/substrate)  (not `substrate-debug`) which seems to be not used for more then 1 year.

@alvicsam what do you think?

File: zombienet/0000_block_building/0000-block-building.toml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[settings]
```
Comment (michalkucharczyk) [inline zombienet/0000_block_building/0000-block-building.toml:1]: `0000-block-building` test is only for testing CI files.
Could be removed.

File: zombienet/0001-basic-warp-sync/0001-README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# Test design
+The `warp-sync` test works on predefined database which is stored in the cloud and
+fetched by the test. `alice` and `bob` nodes are spun up using this database snapshot.
+
+As `warp-sync` requires at least 3 peers, the test spawns the `charlie` node which is `--sync full`.
```
Comment (bkchr) [inline zombienet/0001-basic-warp-sync/0001-README.md:5]: Why not spawn all nodes with the predefined database? I don't get why you let one node sync?

File: zombienet/0001-basic-warp-sync/0001-test-warp-sync.zndsl [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+dave: log line matches "Block history download is complete" within 5 seconds
+
+# db snapshot has 12133 blocks
+dave: reports block height is at least 1 within 5 seconds
+dave: reports block height is at least 12132 within 5 seconds
+dave: reports block height is at least 12133 within 5 seconds
+
+alice: reports block height is at least 12133 within 5 seconds
+bob: reports block height is at least 12133 within 5 seconds
+charlie: reports block height is at least 12133 within 5 seconds
```
Comment (bkchr) [inline zombienet/0001-basic-warp-sync/0001-test-warp-sync.zndsl:30]: These timeouts seem to be too tight. CI systems tend to be overloaded resulting in requiring more time :P

File: zombienet/0001-basic-warp-sync/0001-test-warp-sync.toml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+[relaychain]
+default_image = "docker.io/parity/substrate:master"
+default_command = "/home/miszka/parity/05-zombienet-warp-sync/substrate-wrap.sh"
```
Comment (pepoviola) [inline zombienet/0001-basic-warp-sync/0001-test-warp-sync.toml:3]: Where is this script? Also, the path here should be valid in the image we use.

File: zombienet/0001-basic-warp-sync/0001-README.md (hunk 3 above)
Comment (michalkucharczyk) [reply zombienet/0001-basic-warp-sync/0001-README.md:5]: Did not think of this. Can be done with predefined database also.

File: zombienet/0001-basic-warp-sync/0001-test-warp-sync.toml (hunk 5 above)
Comment (michalkucharczyk) [reply zombienet/0001-basic-warp-sync/0001-test-warp-sync.toml]: fixed in: [0649849](https://github.com/paritytech/substrate/pull/12675/commits/0649849bd2088cfe774b1444b24a0ac4e388e9d2)

File: zombienet/0001-basic-warp-sync/0001-test-warp-sync.zndsl (hunk 4 above)
Comment (michalkucharczyk) [reply zombienet/0001-basic-warp-sync/0001-test-warp-sync.zndsl]: Increased to `30s` in [0649849](https://github.com/paritytech/substrate/pull/12675/commits/0649849bd2088cfe774b1444b24a0ac4e388e9d2)

File: zombienet/0001-basic-warp-sync/0001-README.md (hunk 3 above)
Comment (michalkucharczyk) [reply zombienet/0001-basic-warp-sync/0001-README.md]: Done in: [0649849](https://github.com/paritytech/substrate/pull/12675/commits/0649849bd2088cfe774b1444b24a0ac4e388e9d2)

File: scripts/ci/gitlab/pipeline/publish.yml [hunk 6]
```diff
@@ -2,32 +2,33 @@
...
-    IMAGE_NAME:                    docker.io/parity/$PRODUCT
+    IMAGE_NAME:                    docker.io/$IMAGE_PATH
   before_script:
     - cd ./artifacts/$PRODUCT/
     - VERSION="$(cat ./VERSION)"
     - echo "${PRODUCT} version = ${VERSION}"
     - test -z "${VERSION}" && exit 1
   script:
-    - test "$Docker_Hub_User_Parity" -a "$Docker_Hub_Pass_Parity" ||
+    - test "$COMMON_USER" -a "$COMMON_PASS" ||
```
Comment (alvicsam) [inline scripts/ci/gitlab/pipeline/publish.yml:20]:
```suggestion
    - test "$DOCKER_USER" -a "$DOCKER_PASS" ||
```
Please change "common" to "docker"

**Review by alvicsam — CHANGES_REQUESTED** [review]

File: scripts/ci/gitlab/pipeline/publish.yml (hunk 6 above)
Comment (michalkucharczyk) [reply scripts/ci/gitlab/pipeline/publish.yml]: done in [37b2167](https://github.com/paritytech/substrate/pull/12675/commits/37b216728781a863dd041a4185f7a8540e309f96) (I used `DOCKERIO_USER`).

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -192,6 +194,14 @@ default:
     - if: $CI_COMMIT_REF_NAME == "master"
     - if: $CI_COMMIT_REF_NAME =~ /^v[0-9]+\.[0-9]+.*$/              # i.e. v1.0, v2.1rc1
 
+.build-refs:
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply .gitlab-ci.yml:197]: 8f1cbe4

File: zombienet/0001-basic-warp-sync/0001-generate-warp-sync-database.toml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+[relaychain]
+default_image = "docker.io/paritypr/polkadot-debug:master"
+default_command = "/home/miszka/parity/05-zombienet-warp-sync/substrate-wrap.sh"
```
Comment (skunert) [inline zombienet/0001-basic-warp-sync/0001-generate-warp-sync-database.toml:3]: I think this one slipped through

File: scripts/ci/gitlab/pipeline/publish.yml [hunk 9]
```diff
@@ -2,32 +2,33 @@
...
-    IMAGE_NAME:                    docker.io/parity/$PRODUCT
+    IMAGE_NAME:                    docker.io/$IMAGE_PATH
   before_script:
     - cd ./artifacts/$PRODUCT/
     - VERSION="$(cat ./VERSION)"
     - echo "${PRODUCT} version = ${VERSION}"
     - test -z "${VERSION}" && exit 1
   script:
-    - test "$Docker_Hub_User_Parity" -a "$Docker_Hub_Pass_Parity" ||
+    - test "$DOCKERIO_USER" -a "$DOCKERIO_PASS" ||
```
Comment (alvicsam) [inline scripts/ci/gitlab/pipeline/publish.yml:20]: Please, can you use "DOCKER" so it's consistent to other pipeline configs (for example: https://github.com/paritytech/polkadot/blob/master/scripts/ci/gitlab/pipeline/publish.yml#L22)

**Review by alvicsam — CHANGES_REQUESTED** [review]

File: scripts/ci/gitlab/pipeline/publish.yml (hunk 9 above)
Comment (michalkucharczyk) [reply scripts/ci/gitlab/pipeline/publish.yml]: ok, done: [80de118](https://github.com/paritytech/substrate/pull/12675/commits/80de118a0d16328725a5e17788097e025aa41982)

File: zombienet/0001-basic-warp-sync/0001-generate-warp-sync-database.toml (hunk 8 above)
Comment (michalkucharczyk) [reply zombienet/0001-basic-warp-sync/0001-generate-warp-sync-database.toml]: thx, done in [6ac5d72](https://github.com/paritytech/substrate/pull/12675/commits/6ac5d728f5217c81102de629346f2a16c4c77a5c)

**Review by alvicsam — APPROVED** [review]

File: scripts/ci/gitlab/pipeline/publish.yml (hunk 9 above)
Comment (alvicsam) [reply scripts/ci/gitlab/pipeline/publish.yml]: thanks!

**Review by sergejparity — APPROVED** [review]

**Comment by michalkucharczyk** [comment]
bot rebase

File: zombienet/0000_block_building/0000-block-building.toml (hunk 2 above)
Comment (michalkucharczyk) [reply zombienet/0000_block_building/0000-block-building.toml:1]: Agreed to keep it.

**Review by pepoviola — APPROVED** [review]

**Comment by michalkucharczyk** [comment]
Looks like warp-sync test is working quite stable (I tested 10 executions):
https://gitlab.parity.io/parity/mirrors/substrate/-/pipelines/228196/builds

**Review by skunert — APPROVED** [review]

**Comment by michalkucharczyk** [comment]
Just for the record: 
The `zombienet` testing warp-sync with two validators, and epoch-skipping PR was evaluated here: https://github.com/paritytech/substrate/pull/12751. At the time of writing this comment it is failing.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2080073

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2081963

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
bot rebase

**Review by bkchr — APPROVED** [review]
I find all this `0001-*` naming weird, especially that we also call the files in this way, but that is something we can always fix.

**Comment by michalkucharczyk** [comment]
bot merge

**Pull request merged**